packages/gitbook/src/components/primitives/LoadingStateProvider.tsx

```diff
@@ -0,0 +1,50 @@
+'use client';
+
+import { createContext, useContext, useEffect, useMemo, useState } from 'react';
+
+export const LoadingStateProviderContext = createContext<{
+    bodyLoaded: boolean;
+    setBodyLoaded: (loaded: boolean) => void;
+}>({
+    bodyLoaded: false,
+    setBodyLoaded: () => {},
+});
+
+/**
+ * A provider that tracks the loading state of the body.
+ * This is used to determine when the body has finished loading.
+ * If we need to track more loading states in the future, we can extend this context.
+ */
+export function LoadingStateProvider({
+    children,
+}: {
+    children: React.ReactNode;
+}) {
+    const [bodyLoaded, setBodyLoaded] = useState(false);
+
+    return (
+        <LoadingStateProviderContext.Provider value={{ bodyLoaded, setBodyLoaded }}>
```

Comment:
You can't do this, it'll re-render every component that listens to this

Use `useMemo`

Otherwise you can also use `zustand`

Reply:
It doesn't matter if the client component that uses this context uses `useMemo`.

If you do `value={{ something }}` in aReact context provder, it means every time the provider re-renders, all the clients that calls the associated `useContext` will re-render; because it uses `Object.is` to compare values.

```suggestion
        <LoadingStateProviderContext.Provider value={React.useMemo(() => ({ bodyLoaded, setBodyLoaded }), [bodyLoaded, setBodyLoaded])}>
```